src/Compilers/Core/Portable/Syntax/SyntaxTree.cs

```diff
@@ -369,11 +370,11 @@ internal int GetDisplayLineNumber(TextSpan span)
         /// </summary>
         internal Cci.DebugSourceInfo GetDebugSourceInfo()
         {
-            if (_lazyChecksum.IsDefault)
+            if (RoslynImmutableInterlocked.VolatileRead(ref _lazyChecksum).IsDefault)
```

Comment:
Stumbling on this was frankly a bit shocking. My only conclusion is that `GetDebugSourceInfo` can never be called from threaded code, because no speculation or reordering is required to race here. It just sets the fields in the wrong order!